### ISSUE TITLE
Document notes on core neural network algos including gradient descent, adagrad, auto-differentiation, etc

### ISSUE THREAD

**Issue by cfregly** [fluxcapacitor/pipeline#104]
Notes from the Twitter guy's Automatic Differentiation [video](https://www.youtube.com/watch?v=sq2gPzlrM0g)

Re: O(n^3) when going left to right vs O(n^2) when going right to left. The idea is fairly basic. 
Take for example a "standard" multiplication problem:

2 * 3 * 4 * 5

One way you could approach this is by starting from the left and multiplying left most numbers,
 subbing in their result, and continuing until you have the final answer:

(2 * 3) * 4 * 5
= (6 * 4) * 5
= 24 * 5
= 120

You can also do this from right to left and get the same effect:

2 * 3 * (4 * 5)
= 2 * (3 * 20)
= 2 * 60
= 120

With simple scalar numbers, it doesn't matter which direction you go. However, with matrix math, which matrices/vectors you multiply affect the shape of the remaining matrices. I'll use [MxN] to mean "n by m matrix". Note that a [Mx1] matrix is the same as a vector (just one column). Here's an example matrix multiplication problem:

[200x300] * [300x400] * [400x100] * [400x1]

If you do your matrix math from left to right, notice that each computation is a matrix-matrix computation until the last step:

( [200x300] * [300x400] ) * [400x100] * [400x1]
= ( [200x400] * [400x100] ) * [400x1]
=   [200x100] * [400x1]
=   [200x1]

On the flip side, if you do it in the other direction, each computation is now a matrix-vector computation:

I just noticed that the second to last matrix in the above figures need to be [400x400] instead of [400x100]

Fixed "bad" way of doing it
( [200x300] * [300x400] ) * [400x400] * [400x1]
= ( [200x400] * [400x400] ) * [400x1]
=   [200x400] * [400x1]
=   [200x1]

Right to left "good way":

[200x300] * [300x400] * ( [400x400] * [400x1] )
= [200x300] * ( [300x400] * [400x1] )
= [200x300] * [300x1]
= [200x1]

*************************************************
Each GPU has an associate amount of cores with it- if you look up the model on the NVIDIA website to find it, though I'd bet it's listed in the hardware specification on the computer. I'll check it out when I get back home later today- about to head out hiking.

But in general, just take the matrix and multiply its dimensions- if m*n < cores its saturated

Then there's filling up the GPU ram, which is standard memory allocation math

For combining matrices, I need to look at the SparseTensor implementation in TensorFlow, because when you do that "batching" I drew out, you end up with quadric amounts of zeroes

**Comment by cfregly** [comment]
Here's a presentation that @samjabrahams created:  https://goo.gl/tKOvr7

Here's the meetup where this was presented:  http://www.meetup.com/Advanced-Spark-and-TensorFlow-Meetup/events/231457813/
